File: src/main/java/com/adinsight/agent/loop/AgentLoop.java
package com.adinsight.agent.loop;

import com.adinsight.agent.context.AgentContext;
import com.adinsight.agent.context.ContextBuilder;
import com.adinsight.agent.model.AgentModel;
import com.adinsight.agent.model.AssistantMessage;
import com.adinsight.agent.model.ModelRequest;
import com.adinsight.agent.model.ToolCall;
import com.adinsight.agent.prompt.PromptProvider;
import com.adinsight.agent.session.AgentSession;
import com.adinsight.agent.session.AgentTurn;
import com.adinsight.agent.tools.AgentTool;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;

public class AgentLoop {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final AgentModel model;
    private final Map<String, AgentTool<?>> tools;
    private final PromptProvider promptProvider;
    private final ContextBuilder contextBuilder;
    private final int maxSteps;

    public AgentLoop(
            AgentModel model,
            Map<String, AgentTool<?>> tools,
            PromptProvider promptProvider,
            ContextBuilder contextBuilder,
            int maxSteps
    ) {
        this.model = model;
        this.tools = Map.copyOf(tools);
        this.promptProvider = promptProvider;
        this.contextBuilder = contextBuilder;
        this.maxSteps = maxSteps;
    }

    public CompletableFuture<AgentLoopResult> runTurn(AgentSession session, String userMessage) {
        session.append(AgentTurn.user(userMessage));
        return runIterations(session, 0, null, Map.of());
    }

    private CompletableFuture<AgentLoopResult> runIterations(
            AgentSession session,
            int turnCount,
            String lastModel,
            Map<String, Object> lastUsage
    ) {
        if (turnCount >= maxSteps) {
            return forceSummary(session, turnCount, lastModel, lastUsage);
        }

        ModelRequest request = prepareModelRequest(session, "auto");
        return model.complete(request)
                .thenCompose(response -> {
                    AssistantMessage message = response.message();
                    if (!message.needsFollowUp()) {
                        session.append(AgentTurn.assistant(
                                message.content(),
                                List.of(),
                                response.model(),
                                response.usage()
                        ));
                        return CompletableFuture.completedFuture(new AgentLoopResult(
                                message.content(),
                                response.model(),
                                response.usage(),
                                FinishReason.COMPLETED,
                                turnCount
                        ));
                    }

                    session.append(AgentTurn.assistant(
                            message.content(),
                            message.toolCalls(),
                            response.model(),
                            response.usage()
                    ));

                    return executeTools(session, message.toolCalls())
                            .thenCompose(ignored -> runIterations(
                                    session,
                                    turnCount + 1,
                                    response.model(),
                                    response.usage()
                            ));
                })
                .exceptionally(ex -> {
                    String content = "Model call failed: " + rootMessage(ex);
                    session.append(AgentTurn.assistant(content, List.of(), lastModel, lastUsage));
                    return new AgentLoopResult(content, lastModel, lastUsage, FinishReason.MODEL_ERROR, turnCount);
                });
    }

    private ModelRequest prepareModelRequest(AgentSession session, String toolChoice) {
        List<Map<String, Object>> messages = buildOpenAiMessages(session);
        List<Map<String, Object>> toolDefs = tools.values().stream()
                .map(AgentTool::toOpenAiToolDef)
                .toList();

        return new ModelRequest(
                messages,
                toolDefs,
                toolChoice,
                session.numericUserId().isPresent() ? session.numericUserId().getAsLong() : null,
                session.numericSessionId().isPresent() ? session.numericSessionId().getAsLong() : null
        );
    }

    private CompletableFuture<Void> executeTools(AgentSession session, List<ToolCall> toolCalls) {
        CompletableFuture<Void> chain = CompletableFuture.completedFuture(null);
        for (ToolCall toolCall : toolCalls) {
            chain = chain.thenCompose(ignored -> executeTool(session, toolCall));
        }
        return chain;
    }

    private CompletableFuture<Void> executeTool(AgentSession session, ToolCall toolCall) {
        AgentTool<?> tool = tools.get(toolCall.name());
        Map<String, Object> args = parseArgs(toolCall.argumentsJson());
        CompletableFuture<Map<String, Object>> resultFuture = tool == null
                ? CompletableFuture.completedFuture(errorResult("Unknown tool: " + toolCall.name()))
                : tool.call(args, session).exceptionally(ex -> errorResult(rootMessage(ex)));

        return resultFuture.thenAccept(result -> session.append(AgentTurn.tool(
                toolCall.id(),
                toolCall.name(),
                args,
                summarizeToolResult(result)
        )));
    }

    private CompletableFuture<AgentLoopResult> forceSummary(
            AgentSession session,
            int turnCount,
            String lastModel,
            Map<String, Object> lastUsage
    ) {
        return model.complete(prepareModelRequest(session, "none"))
                .thenApply(response -> {
                    String content = response.message().content().isBlank()
                            ? "Too many steps; stopped. Please narrow the request and try again."
                            : response.message().content();
                    session.append(AgentTurn.assistant(content, List.of(), response.model(), response.usage()));
                    return new AgentLoopResult(
                            content,
                            response.model() == null ? lastModel : response.model(),
                            response.usage().isEmpty() ? lastUsage : response.usage(),
                            FinishReason.FORCED_SUMMARY,
                            turnCount
                    );
                });
    }

    private List<Map<String, Object>> buildOpenAiMessages(AgentSession session) {
        AgentContext context = contextBuilder.build(session);
        List<Map<String, Object>> messages = new ArrayList<>();
        messages.add(Map.of("role", "system", "content", promptProvider.systemPrompt()));
        messages.add(Map.of("role", "system", "content", runtimeContext(context)));
        messages.addAll(toOpenAiMessages(context.messages()));
        return messages;
    }

    private String runtimeContext(AgentContext context) {
        try {
            return "Runtime context:\n" + MAPPER.writeValueAsString(Map.of(
                    "session_id", context.sessionId(),
                    "state", context.state()
            ));
        } catch (JsonProcessingException e) {
            return "Runtime context: {}";
        }
    }

    private static List<Map<String, Object>> toOpenAiMessages(List<Map<String, Object>> items) {
        List<Map<String, Object>> messages = new ArrayList<>();
        Set<String> validToolCallIds = new HashSet<>();

        for (Map<String, Object> item : items) {
            String role = String.valueOf(item.get("role"));
            if ("assistant".equals(role)) {
                Map<String, Object> message = new LinkedHashMap<>();
                message.put("role", "assistant");
                message.put("content", item.getOrDefault("content", ""));
                Object toolCalls = item.get("tool_calls");
                if (toolCalls instanceof List<?> calls && !calls.isEmpty()) {
                    message.put("tool_calls", toolCalls);
                    for (Object call : calls) {
                        if (call instanceof Map<?, ?> callMap && callMap.get("id") != null) {
                            validToolCallIds.add(String.valueOf(callMap.get("id")));
                        }
                    }
                }
                messages.add(message);
            } else if ("tool".equals(role)) {
                Object toolCallId = item.get("tool_call_id");
                if (toolCallId != null && validToolCallIds.contains(String.valueOf(toolCallId))) {
                    Map<String, Object> message = new LinkedHashMap<>();
                    message.put("role", "tool");
                    message.put("tool_call_id", toolCallId);
                    message.put("name", item.get("tool_name"));
                    message.put("content", item.getOrDefault("content", ""));
                    messages.add(message);
                }
            } else if ("user".equals(role) || "system".equals(role)) {
                messages.add(Map.of("role", role, "content", item.getOrDefault("content", "")));
            }
        }

        return messages;
    }

    private static Map<String, Object> parseArgs(String rawArgs) {
        try {
            Object value = MAPPER.readValue(rawArgs == null || rawArgs.isBlank() ? "{}" : rawArgs, Object.class);
            if (value instanceof Map<?, ?> map) {
                Map<String, Object> result = new LinkedHashMap<>();
                map.forEach((key, item) -> result.put(String.valueOf(key), item));
                return result;
            }
        } catch (Exception ignored) {
        }
        return Map.of();
    }

    private static Map<String, Object> errorResult(String error) {
        return Map.of("error", error);
    }

    private static String summarizeToolResult(Map<String, Object> result) {
        try {
            String text = MAPPER.writeValueAsString(result);
            int maxChars = 1200;
            return text.length() <= maxChars ? text : text.substring(0, maxChars) + "...[truncated]";
        } catch (JsonProcessingException e) {
            return String.valueOf(result);
        }
    }

    private static String rootMessage(Throwable throwable) {
        Throwable current = throwable;
        while (current.getCause() != null) {
            current = current.getCause();
        }
        return current.getMessage() == null ? current.getClass().getSimpleName() : current.getMessage();
    }
}

File: src/loop/agent-loop.ts
import type { ContextBuilder } from "../context/context-builder.js";
import type { AgentModel } from "../model/agent-model.js";
import { needsFollowUp } from "../model/assistant-message.js";
import type { ModelRequest, ToolChoice } from "../model/model-request.js";
import {
  buildOpenAiMessages,
  toOpenAiToolDefinition
} from "../model/openai-adapter.js";
import type { ToolCall } from "../model/tool-call.js";
import { AgentSession } from "../session/agent-session.js";
import { assistantTurn, toolTurn } from "../session/agent-turn.js";
import { callTool, type AgentTool, type ShapedToolResult, type ToolArgs } from "../tools/agent-tool.js";
import type { AgentLoopResult, Usage } from "./agent-loop-result.js";
import { FinishReason } from "./finish-reason.js";

const MAX_TOOL_RESULT_CHARS = 1_200;

export class AgentLoop {
  private readonly tools: ReadonlyMap<string, AgentTool>;

  constructor(
    private readonly model: AgentModel,
    tools: ReadonlyMap<string, AgentTool> | Readonly<Record<string, AgentTool>>,
    private readonly systemPrompt: string,
    private readonly contextBuilder: ContextBuilder,
    private readonly maxSteps: number
  ) {
    this.tools = tools instanceof Map ? new Map(tools) : new Map(Object.entries(tools));
  }

  async runTurn(session: AgentSession, userMessage: string): Promise<AgentLoopResult> {
    session.appendUser(userMessage);
    return this.runIterations(session, 0, undefined, {});
  }

  private async runIterations(
    session: AgentSession,
    turnCount: number,
    lastModel: string | undefined,
    lastUsage: Usage
  ): Promise<AgentLoopResult> {
    if (turnCount >= this.maxSteps) {
      return this.forceSummary(session, turnCount, lastModel, lastUsage);
    }

    try {
      const response = await this.model.complete(this.prepareModelRequest(session, "auto"));
      const message = response.message;

      if (!needsFollowUp(message)) {
        session.append(assistantTurn({
          content: message.content,
          model: response.model,
          usage: response.usage
        }));
        return {
          content: message.content,
          model: response.model,
          usage: response.usage,
          finishReason: FinishReason.Completed,
          turnCount
        };
      }

      session.append(assistantTurn({
        content: message.content,
        toolCalls: message.toolCalls,
        model: response.model,
        usage: response.usage
      }));

      await this.executeTools(session, message.toolCalls);
      return this.runIterations(session, turnCount + 1, response.model, response.usage);
    } catch (error) {
      const content = `Model call failed: ${rootMessage(error)}`;
      session.append(assistantTurn({
        content,
        model: lastModel,
        usage: lastUsage
      }));
      return {
        content,
        model: lastModel,
        usage: lastUsage,
        finishReason: FinishReason.ModelError,
        turnCount
      };
    }
  }

  private prepareModelRequest(session: AgentSession, toolChoice: ToolChoice): ModelRequest {
    const context = this.contextBuilder.build(session);

    return {
      messages: buildOpenAiMessages({
        systemPrompt: this.systemPrompt,
        context
      }),
      tools: [...this.tools.values()].map(toOpenAiToolDefinition),
      toolChoice,
      userId: session.numericUserId(),
      sessionId: session.numericSessionId()
    };
  }

  private async executeTools(session: AgentSession, toolCalls: readonly ToolCall[]): Promise<void> {
    const turns = await Promise.all(toolCalls.map((toolCall) => this.executeTool(session, toolCall)));
    for (const turn of turns) {
      session.append(turn);
    }
  }

  private async executeTool(session: AgentSession, toolCall: ToolCall) {
    const args = parseArgs(toolCall.argumentsJson);
    const tool = this.tools.get(toolCall.name);
    const result = tool == null
      ? {
          ok: false,
          error: `Unknown tool: ${toolCall.name}`,
          metadata: {}
        } satisfies ShapedToolResult
      : await callTool(tool, args, session).catch((error: unknown) => ({
          ok: false,
          error: rootMessage(error),
          metadata: {}
        }) satisfies ShapedToolResult);

    return toolTurn({
      toolCallId: toolCall.id,
      toolName: toolCall.name,
      toolArgs: args,
      content: summarizeToolResult(result)
    });
  }

  private async forceSummary(
    session: AgentSession,
    turnCount: number,
    lastModel: string | undefined,
    lastUsage: Usage
  ): Promise<AgentLoopResult> {
    const response = await this.model.complete(this.prepareModelRequest(session, "none"));
    const content = response.message.content.trim() === ""
      ? "Too many steps; stopped. Please narrow the request and try again."
      : response.message.content;
    const model = response.model ?? lastModel;
    const usage = Object.keys(response.usage).length === 0 ? lastUsage : response.usage;

    session.append(assistantTurn({
      content,
      model,
      usage
    }));

    return {
      content,
      model,
      usage,
      finishReason: FinishReason.ForcedSummary,
      turnCount
    };
  }
}

function parseArgs(rawArgs: string): ToolArgs {
  try {
    const value: unknown = JSON.parse(rawArgs.trim() === "" ? "{}" : rawArgs);
    if (value != null && typeof value === "object" && !Array.isArray(value)) {
      return { ...(value as Record<string, unknown>) };
    }
  } catch (error) {
    if (!(error instanceof SyntaxError)) {
      throw error;
    }
  }

  return {};
}

function summarizeToolResult(result: ShapedToolResult): string {
  const text = JSON.stringify(result);
  return text.length <= MAX_TOOL_RESULT_CHARS
    ? text
    : `${text.slice(0, MAX_TOOL_RESULT_CHARS)}...[truncated]`;
}

function rootMessage(error: unknown): string {
  if (!(error instanceof Error)) {
    return String(error);
  }

  let current: Error = error;
  while (current.cause instanceof Error) {
    current = current.cause;
  }

  return current.message === "" ? current.name : current.message;
}
